### ISSUE TITLE
Asp.Net Core Dependency Injection problems

### ISSUE THREAD

**Issue by raffaeledoti** [thepirat000/Audit.NET#387]
I'm trying to inject , at startup , all dependencies into my custom data provider in order to save audit event into a custom data table. My goal is to have 2 new columns to save user id and user tenant but the  problem is that i cannot inject my data provider dependencies at startup because Ioc service provider cannot resolve scoped dependencies. Is there any way to inject the custom data provider instead of manually instanciate it?

**To Reproduce**

My Startup.cs code :

```
#region AuditConfiguration
            //configure database output
            Audit.Core.Configuration.Setup().UseCustomProvider(app.ApplicationServices.GetRequiredService<CustomDataProvider>());
            //add tenant and user id to all intercepted http request
            Audit.Core.Configuration.AddCustomAction(ActionType.OnEventSaving, scope =>
            {
                //get claims provider 
                var claimsProvider = app.ApplicationServices.GetRequiredService<IClaimsProvider>();
                // Add new fields
                scope.Event.CustomFields["Tenant"] = claimsProvider.GetUserTenant();
                scope.Event.CustomFields["UserId"] = claimsProvider.GetUserId().ToString();
            });
            //determine which http request must be logged
            app.UseAuditMiddleware(options => options.FilterByRequest(httpReq => httpReq.Method.ToUpper().Equals("POST") || httpReq.Method.ToUpper().Equals("PUT") || httpReq.Method.ToUpper().Equals("DELETE")));
#endregion 
```

**Expected behavior**

I would expect that my dependencies will successfully injected in order to save data to a custom table

**Libraries** 
 - Audit.WebApi

**Target .NET framework:**
 - .NET Core 3.1

**Comment by thepirat000** [comment]
I'll check if it can provide a way to specify a builder instead of an instance.

Will something like this work for you?

```c#
Audit.Core.Configuration.Setup()
	.UseBuilder(() => app.ApplicationServices.GetRequiredService<CustomDataProvider>())
```

**Comment by thepirat000** [comment]
This was implemented on the new version (17.0.0), please upgrade your references and re-test.

Now the **data provider** can be set as a factory which will be called when an AuditScope is created, or when the property getter `Audit.Core.Configuration.DataProvider` is accessed.

for example:

```c#
Audit.Core.Configuration.Setup()
	.UseFactory(() => app.ApplicationServices.GetRequiredService<CustomDataProvider>());
```

Or:

```c#
Audit.Core.Configuration.DataProviderFactory = () => app.ApplicationServices.GetRequiredService<CustomDataProvider>();
```

**Comment by raffaeledoti** [comment]
Thank you I will try asap!

**Comment by prawana-perera** [comment]
Bit late, but I ran across same issue. Our system is a multi-tenant app server with single tenant databases. So every web request we create a new request lifetime scoped (via AutoFac), this included the DbContext which was hooked up to the user's organisation database.

We are writing the audit to one Cosmos DB and we are separating each organisation using containers, hence we needed the Container to be dynamic. Also one additional requirement was that we also had to support background jobs (via HangFire running on same app server).

We tried the `UseFactory` but ran into issues as the DataProvider had to be InstancePerRequest (for web requests) or InstancePerLifetimeScope (for the scope of the background job). And when setting up Audit.Net Configuration (once via Startup) and resolving the dependency it did no have access to the child scopes.

We needed the userId and the orgId which were in the HttpContext but this was not possible in background jobs.

We moved these to a different class which had this state, and we had 2 implementations of it, one for web request (the values were pulled from HttpContext) and one for background jobs (initilased when the job ran). This object was under DI and was created for the required lifetime scope. 

I ended up changing AuditDataProvider in the DBContext, i.e.

```        
public class SomeAppEntities : AuditDbContext, ISomeAppEntities
{
  public SomeAppEntities(IDbConnectionProvider dbConnectionProvider, ISomeAppConfigurationProvider someAppConfigurationProvider) : base(dbConnectionProvider.GetConnectionString())
          {
              var endpoint = someAppConfigurationProvider.GetAuditDatabaseConnectionEndpoint();
              var authKey = someAppConfigurationProvider.GetAuditDatabaseAuthorizationKey();
              var container = someAppConfigurationProvider.GetCustomerOrgId();
  
              AuditDataProvider = new AzureCosmosDataProvider()
              {
                  Endpoint = endpoint,
                  AuthKey = authKey,
                  Database = AuditingDatabaseConstants.DatabaseName,
                  Container = container,
                  ConnectionPolicy = new ConnectionPolicy() 
                  {
                      ConnectionMode = ConnectionMode.Gateway
                  }
              };
          }
...
}
```

**Comment by c5racing** [comment]
I like setting the DataProvider in the DBContext so I can use DI to get user specific information into the audit logs.  The only problem is while doing this, some logs, specifically, ones generated by:

`optionsBuilder.AddInterceptors(new AuditCommandInterceptor());`

 are written to json files.

In my startup.cs file, I have

`Audit.Core.Configuration.Setup();`

In my Dbcontext.cs file, I have:

```
        AuditDataProvider = new PostgreSqlDataProvider()
            { 
                ConnectionString = auditSettings.Value.ConnectionString,
                TableName = "audit_event",
                IdColumnName = "id",
                DataColumnName = "data",
                DataType = Audit.PostgreSql.Configuration.DataType.JSONB.ToString(),
                LastUpdatedDateColumnName = "created_date",
            };
```

Why are the interceptors not using the AuditDataProvider specified?

**Comment by thepirat000** [comment]
The command interceptor does not depend on the `IAuditDbContext`. You can attach the command interceptor to any `DbContext` regardless of whether it inherits from `AuditDbContext`. 

That's the main reason why the `AuditDataProvider` you've set on your `AuditDbContext` is not taken into account for the command interceptor.

In the current version, the command interceptor uses the global data provider from the static property `Audit.Core.Configuration.AuditDataProvider`.

You can set the global data provider by calling `Audit.Core.Configuration.Setup().UsePostgreSql(...);`

By default, the global data provider is a FileDataProvider that saves the events to .json files in the current directory.

Note that calling only `Audit.Core.Configuration.Setup();` will have no effect.

Anyway I think the interceptor can be changed to use the `AuditDataProvider` set in the `DbContext`, when the `DbContext` **is** an `AuditDbContext`.

**Comment by c5racing** [comment]
If the above could be accomplished, that would be great.  Based on the above, I don't have control to change the connection string per Tenent without this.

**Comment by thepirat000** [comment]
The fix will be out soon

Sent from Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Comment by thepirat000** [comment]
This was fixed in version 20.2.2, please upgrade your references and re-test.

Now the interceptor will use the data provider defined in the property `AuditDataProvider` from your `AuditDbContext`

**Comment by c5racing** [comment]
Thank you @thepirat000!  The fix works perfect!
